common/flatpak-dir.c

```diff
@@ -7973,50 +8023,64 @@ find_matching_refs (GHashTable *refs,
   g_hash_table_iter_init (&hash_iter, refs);
   while (g_hash_table_iter_next (&hash_iter, &key, NULL))
     {
-      g_autofree char *ref = NULL;
-      g_auto(GStrv) parts = NULL;
-      gboolean is_app, is_runtime;
       FlatpakCollectionRef *coll_ref = key;
 
-      /* Unprefix any remote name if needed */
-      ostree_parse_refspec (coll_ref->ref_name, NULL, &ref, NULL);
-      if (ref == NULL)
-        continue;
+      add_ref_if_match (coll_ref, opt_name, opt_branch, opt_arch, opt_collection_id, kinds, flags, matched_refs);
+    }
 
-      is_app = g_str_has_prefix (ref, "app/");
-      is_runtime = g_str_has_prefix (ref, "runtime/");
+  return g_steal_pointer (&matched_refs);
+}
 
-      if ((!(kinds & FLATPAK_KINDS_APP) && is_app) ||
-          (!(kinds & FLATPAK_KINDS_RUNTIME) && is_runtime) ||
-          (!is_app && !is_runtime))
-        continue;
+/* This is similar to find_matching_refs but uses xa.cache data instead */
+static GPtrArray *
+find_matching_refs_from_cache (GVariant     *xa_cache,
+                               const char   *cache_collection_id,
+                               const char   *opt_name,
+                               const char   *opt_branch,
+                               const char   *opt_arch,
+                               FlatpakKinds  kinds,
+                               FindMatchingRefsFlags flags,
+                               GError      **error)
+{
+  g_autoptr(GPtrArray) matched_refs = NULL;
+  g_autoptr(GVariant) cache = NULL;
+  g_autoptr(GError) local_error = NULL;
+  gsize i, n;
 
-      parts = flatpak_decompose_ref (ref, NULL);
-      if (parts == NULL)
-        continue;
+  if (opt_name && !flatpak_is_valid_name (opt_name, &local_error))
+    {
+      flatpak_fail (error, "'%s' is not a valid name: %s", opt_name, local_error->message);
+      return NULL;
+    }
 
-      if (opt_name != NULL && strcmp (opt_name, parts[1]) != 0)
-        continue;
+  if (opt_branch && !flatpak_is_valid_branch (opt_branch, &local_error))
+    {
+      flatpak_fail (error, "'%s' is not a valid branch name: %s", opt_branch, local_error->message);
+      return NULL;
+    }
 
-      if (!g_strv_contains (arches, parts[2]))
-        continue;
+  matched_refs = g_ptr_array_new_with_free_func (g_free);
 
-      if (opt_branch != NULL && strcmp (opt_branch, parts[3]) != 0)
-        continue;
+  cache = g_variant_get_child_value (xa_cache, 0);
+  n = g_variant_n_children (cache);
+  for (i = 0; i < n; i++)
+    {
+      g_autoptr(GVariant) child = NULL;
+      g_autoptr(GVariant) cur_v = NULL;
+      g_autoptr(FlatpakCollectionRef) coll_ref = NULL;
+      const char *ref;
 
-      if (opt_collection_id != NULL && strcmp (opt_collection_id, coll_ref->collection_id))
-        continue;
+      child = g_variant_get_child_value (cache, i);
+      cur_v = g_variant_get_child_value (child, 0);
+      ref = g_variant_get_data (cur_v);
```

Comment:
I commented this before, but I believe it would be best to use the typed method to get this data; e.g. `g_variant_get_string()` or `g_variant_get_bytestring()` (not sure what the type of this variant is).